tunix/rl/agentic/agents/tool_agent.py

```diff
@@ -0,0 +1,156 @@
+from tunix.rl.agentic.parser.tool_parser.tool_parser_registry import get_tool_parser
+from tunix.rl.agentic.parser.tool_parser.tool_parser_base import ToolParser
+
+logger = logging.getLogger(__name__)
+
+
+class ToolAgent(LLMBaseAgent):
+    """
+    An Agent implementation that supports tool usage, conforming to the LLMBaseAgent abstract interface.
+    Supports injecting tool_map and parsing with a specified parser.
+    """
+
+    def __init__(
+        self,
+        system_prompt: str,
+        parser_name: str = "qwen",
+        tool_map: dict[str, type[BaseTool]] | None = None,
+    ):
+        self.system_prompt = system_prompt
+
+        # Tool manager (Router)
+        self.tool_manager = ToolManager(tool_map=tool_map or {})
+
+        # Parser (converts LLM responses to tool calls)
+        parser_cls: type[ToolParser] = get_tool_parser(parser_name)
+        self.tool_parser = parser_cls()
+        
+        # Build tools prompt: inject JSON Schema
+        tools_json = json.dumps(self.tool_manager.json, indent=2)
+        self.tools_prompt = self.tool_parser.get_tool_prompt(tools_json)
+        
+        # Internal state
+        self._trajectory = Trajectory()
+        self._messages: list[dict[str, Any]] = []
+        self._obs_cache = None  # Caches the last observation
+
+        self.reset()
+
+    # ─────────────────────────────────────────────────────────────
+    # Property Interfaces
+    # ─────────────────────────────────────────────────────────────
+
+    @property
+    def chat_completions(self) -> list[dict[str, str]]:
+        return self._messages
+
+    @property
+    def trajectory(self) -> Trajectory:
+        return self._trajectory
+
+    # ─────────────────────────────────────────────────────────────
+    # Interaction with Environment
+    # ─────────────────────────────────────────────────────────────
+
+    def update_from_env(self, observation: Any, reward: float, done: bool, info: dict, **kwargs):
+        """Write observation / reward and update context"""
+        step = self.get_current_state()
+        if step:
+            step.observation = observation
+            step.reward = reward
+            step.done = done
+            step.info = info or {}
+
+        # Cache the observation for next round of generation
+        self._obs_cache = observation
+
+        # Convert to messages
+        if isinstance(observation, dict):
+            if "tool_outputs" in observation:
+                for call_id, output in observation["tool_outputs"].items():
+                    self._messages.append({
+                        "role": "user",
+                        "tool_call_id": call_id,
+                        "content": "Tool returned result: " + output,
+                    })
+            elif "question" in observation:
+                self._messages.append({
+                    "role": "user",
+                    "content": observation["question"],
+                })
```

Comment:
what does it mean neither tool_outputs nor question exists, is that a valid case or should throw some exception?